test/request-response-channel.cc

```diff
@@ -0,0 +1,121 @@
+    return this->data == response.data;
+  }
+
+  std::string data{};
+};
+
+TEST(RequestResponseChannel, BunchOfRequests) {
+  RequestResponseChannel<Request, Response> channel;
+
+  auto operation = [&]() {
+    return DoAll(
+        channel.Read()
+            >> Map([&](Request&& request) {
+                Response response{"response for " + request.data};
+                return channel.Respond(std::move(response));
+              })
+            >> Loop<void>()
+                   .context(0)
+                   .body([](auto&& context, auto& stream) {
+                     if (context < 3) {
+                       ++context;
+                       stream.Next();
+                     }
+                     stream.Done();
+                   })
+                   .ended([](auto&& context, auto& k) {
+                     k.Start();
+                   }),
+        Then([&]() {
+          Request request{"request1"};
+          return channel.Request(std::move(request))
+              >> Then([](std::optional<Response>&& response) {
+                   EXPECT_EQ(response->data, "response for request1");
+                 });
+        }),
+        Then([&]() {
+          Request request{"request2"};
+          return channel.Request(std::move(request))
+              >> Then([](std::optional<Response>&& response) {
+                   EXPECT_EQ(response->data, "response for request2");
+                 });
+        }),
+        Then([&]() {
+          Request request{"request3"};
+          return channel.Request(std::move(request))
+              >> Then([](std::optional<Response>&& response) {
+                   EXPECT_EQ(response->data, "response for request3");
+                 });
+        }));
+  };
+
+  EXPECT_EQ(
+      *operation(),
+      std::make_tuple(
+          std::monostate{},
+          std::monostate{},
+          std::monostate{},
+          std::monostate{}));
+}
+
+TEST(RequestResponseChannel, ReadBatch) {
+  RequestResponseChannel<Request, Response> channel;
+
+  auto read = [&]() {
+    return channel.ReadBatch()
+        >> Then([&](std::optional<std::deque<Request>>&& requests) {
+             std::deque<Response> result;
+             for (const Request& request : *requests) {
+               result.push_back(Response{"response for " + request.data});
+             }
+             return channel.RespondBatch(std::move(result));
+           });
+  };
+
+  auto [future1, write1] =
+      PromisifyForTest(channel.Request(Request{"request1"}));
+
+  write1.Start();
+
+  auto [future2, write2] =
+      PromisifyForTest(channel.Request(Request{"request2"}));
+
+  write2.Start();
+
+  *read();
+
+  std::optional<Response> response1 = future1.get();
+  std::optional<Response> response2 = future2.get();
+
+  ASSERT_TRUE(response1);
+  ASSERT_TRUE(response2);
+
+  EXPECT_EQ(response1.value(), Response{"response for request1"});
+  EXPECT_EQ(response2.value(), Response{"response for request2"});
```

Comment:
This is fine, but we can remove the extra `operator==` above and just do:
```suggestion
  EXPECT_EQ(response1->data, "response for request1");
  EXPECT_EQ(response2->data, "response for request2");
```